pkg/k8s/rule_translate.go

```diff
@@ -0,0 +1,180 @@
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+var _ policy.Translator = RuleTranslator{}
+
+// RuleTranslator implements pkg/policy.Translator interface
+// Translate populates/depopulates given rule with ToCIDR rules
+// Based on provided service/endpoint
+type RuleTranslator struct {
+	Service  types.K8sServiceNamespace
+	Endpoint types.K8sServiceEndpoint
+	Revert   bool
+}
+
+// Translate calls TranslateEgress on all r.Egress rules
+func (k RuleTranslator) Translate(r *api.Rule) error {
+	for egressIndex := range r.Egress {
+		err := k.TranslateEgress(&r.Egress[egressIndex])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TranslateEgress populates/depopulates egress rules with ToCIDR entries based
+// on toService entries
+func (k RuleTranslator) TranslateEgress(r *api.EgressRule) error {
+	err := k.depopulateEgress(r)
+	if err != nil {
+		return err
+	}
+	if !k.Revert {
+		err := k.populateEgress(r)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) populateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := generateToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+func (k RuleTranslator) depopulateEgress(r *api.EgressRule) error {
+	for _, service := range r.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == api.K8sServiceNamespace(k.Service) {
+			if err := deleteToCidrFromEndpoint(r, k.Endpoint); err != nil {
+				return err
+			}
+			// TODO: generateToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+// generateToCidrFromEndpoint takes an egress rule and populates it with
+// ToCIDR rules based on provided endpoint object
+func generateToCidrFromEndpoint(
+	egress *api.EgressRule, endpoint types.K8sServiceEndpoint) error {
+
+	// This will generate one-address CIDRs consisting of endpoint backend ip
+	mask := net.CIDRMask(128, 128)
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+		if epIP == nil {
+			return fmt.Errorf("Unable to parse ip: %s", ip)
+		}
+
+		found := false
+		for _, c := range egress.ToCIDR {
+			_, cidr, err := net.ParseCIDR(string(c))
+			if err != nil {
+				return err
+			}
+			if cidr.Contains(epIP) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, api.CIDR(cidr.String()))
+		}
+	}
+	return nil
+}
+
+// deleteToCidrFromEndpoint takes an egress rule and removes
+// ToCIDR rules matching endpoint
+func deleteToCidrFromEndpoint(
```

Comment:
Is this method called anywhere?
It doesn't seem to work, cf. my other comment below.
If it's not used anywhere, remove it.